arangod/Cluster/AgencyCache.cpp

```diff
@@ -470,3 +556,111 @@ void AgencyCache::invokeAllCallbacks() const {
+      AgencyCommHelper::path(PLAN_COLLECTIONS) + "/",
+      AgencyCommHelper::path(PLAN_DATABASES) + "/",
+      AgencyCommHelper::path(PLAN_VIEWS) + "/"});
+  static std::vector<std::string> const currentGoodies ({
+      AgencyCommHelper::path(CURRENT_COLLECTIONS) + "/",
+      AgencyCommHelper::path(CURRENT_DATABASES) + "/"});
+  
+  bool get_rest = false;
+  std::unordered_map<std::string, query_t> db_res;
+  query_t rest_res = nullptr;
+
+  decltype(_planChanges) const& changes = (what == PLAN) ? _planChanges : _currentChanges;
+  std::vector<std::string> const& goodies = (what == PLAN) ? planGoodies : currentGoodies;
+  
+  std::unordered_set<std::string> databases;
+  std::shared_lock_guard g(_storeLock);
+
+  if (last < _lastSnapshot) {
+    get_rest = true;
+    auto keys = _readDB.nodePtr(AgencyCommHelper::path(what) + "/" + DATABASES)->keys();
+    for (auto const& i : keys) {
+      databases.emplace(i);
+    }
+  } else {
+    auto it = changes.lower_bound(last+1); 
+    if (it != changes.end()) {
+      for (; it != changes.end(); ++it) {
+        if (it->second.empty()) { // Need to get rest
+          get_rest = true;
+        }
+        databases.emplace(it->second);
+      }
+      LOG_TOPIC("d5743", TRACE, Logger::CLUSTER)
+        << "collecting " << databases << " from agency cache";
+    } else {
+      LOG_TOPIC("d5734", DEBUG, Logger::CLUSTER) << "no changed databases since " << last;
+      return change_set_t(_commitIndex, std::move(db_res), std::move(rest_res));
+    }
+  }
+  
+  if (databases.empty()) {
+    return change_set_t(_commitIndex, std::move(db_res), std::move(rest_res));
+  }
+
+  auto query = std::make_shared<arangodb::velocypack::Builder>();
+  {
+    for (auto const& i : databases) {
+      if (!i.empty()) { // database
+        auto query = std::make_shared<arangodb::velocypack::Builder>();
+        { VPackArrayBuilder outer(query.get());
+          { VPackArrayBuilder inner(query.get());
+            for (auto const& g : goodies) { // Get goodies for database
+              query->add(VPackValue(g + i));
+            }
+          }}
+        auto [entry,created] = db_res.try_emplace(i, std::make_shared<VPackBuilder>());
+        if (created) {
+          _readDB.read(query, entry->second);
+        } else {
+          LOG_TOPIC("31ae3", ERR, Logger::CLUSTER)
+            << "Failed to communicate updated database " << i
+            << " in AgencyCache with maintenance.";
+          FATAL_ERROR_EXIT();
+        }
+      }
+    }
+  }
+
+  if (get_rest) { // All the rest, i.e. All keys excluding the usual suspects
+    static std::vector<std::string> const exc {
+      "Analyzers", "Collections", "Databases", "Views"};
+    auto keys = _readDB.nodePtr(AgencyCommHelper::path(what))->keys();
+    keys.erase(
+      std::remove_if(
+        std::begin(keys), std::end(keys),
+        [&] (auto x) {
+          return std::binary_search(std::begin(exc), std::end(exc),x);}),
```

Comment:
Probably not worthwhille in a list of 4 things.

Reply:
it's most efficient in any case.

Reply:
No problem, let's leave it. But I strongly believe that a linear search is faster for 4 elements. Never mind.